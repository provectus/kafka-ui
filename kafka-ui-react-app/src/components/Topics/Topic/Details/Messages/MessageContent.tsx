import React from 'react';

import FullMessage from './FullMessage';

interface MessageContentProps {
  message: string;
}

const MessageContent: React.FC<MessageContentProps> = ({ message }) => {
  const [isFolded, setIsFolded] = React.useState(message.length > 250);

  return (
    <div className="is-flex is-flex-direction-column">
      {isFolded ? (
        <p className="has-content-overflow-ellipsis">
          {`${message.slice(0, 250)}...`}
        </p>
      ) : (
        <FullMessage message={message} />
      )}
      {isFolded && (
        <button
          type="button"
          className="button is-small mt-2"
          onClick={() => setIsFolded((state) => !state)}
          title="Expand to JSON"
        >
          <span className="icon is-small">
            <i className="fas fa-chevron-down" />
          </span>
        </button>
      )}
    </div>
  );
};

export default MessageContent;
