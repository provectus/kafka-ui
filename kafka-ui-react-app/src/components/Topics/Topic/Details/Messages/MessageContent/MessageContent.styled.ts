import styled from 'styled-components';
import { Colors } from 'theme/theme';

export const MessageContentWrapper = styled.tr`
  background-color: ${Colors.neutral[5]};
  & > td {
    padding: 16px;
    &:first-child {
      padding-right: 1px;
    }
    &:last-child {
      padding-left: 1px;
    }
  }
`;

export const StyledSection = styled.div`
  padding: 0 16px;
  display: flex;
  gap: 1px;
  align-items: stretch;
`;

export const ContentBox = styled.div`
  background-color: white;
  padding: 24px;
  border-radius: 8px 0 0 8px;
  flex-grow: 3;
  & nav {
    padding-bottom: 16px;
  }
`;

export const MetadataWrapper = styled.div`
  background-color: white;
  padding: 24px;
  border-radius: 0 8px 8px 0;
  flex-grow: 1;
  display: flex;
  flex-direction: column;
  gap: 16px;
  min-width: 400px;
`;

export const Metadata = styled.span`
  display: flex;
  gap: 16px;
`;

export const MetadataLabel = styled.p`
  color: ${Colors.neutral[50]};
  font-size: 14px;
  width: 80px;
`;

export const MetadataValue = styled.p`
  color: ${Colors.neutral[80]};
  font-size: 14px;
`;

export const MetadataMeta = styled.p`
  color: ${Colors.neutral[30]};
  font-size: 12px;
`;
