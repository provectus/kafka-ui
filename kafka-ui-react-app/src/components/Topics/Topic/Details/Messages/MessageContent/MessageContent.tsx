import { TopicMessageTimestampTypeEnum } from 'generated-sources';
import React from 'react';
import JSONViewer from 'components/common/JSONViewer/JSONViewer';
import SecondaryTabsStyles from 'components/common/Tabs/SecondaryTabs.styled';
import { isObject } from 'lodash';
import BytesFormatted from 'components/common/BytesFormatted/BytesFormatted';

import {
  ContentBox,
  JSONViewerWrapper,
  MessageContentWrapper,
  Metadata,
  MetadataLabel,
  MetadataMeta,
  MetadataValue,
  MetadataWrapper,
} from './MessageContent.styled';

type Tab = 'key' | 'content' | 'headers';

export interface MessageContentProps {
  messageKey?: string;
  messageContent?: string;
  headers?: { [key: string]: string | undefined };
  timestamp?: Date;
  timestampType?: TopicMessageTimestampTypeEnum;
}

const MessageContent: React.FC<MessageContentProps> = ({
  messageKey,
  messageContent,
  headers,
  timestamp,
  timestampType,
}) => {
  const [activeTab, setActiveTab] = React.useState<Tab>('content');
  const activeTabContent = () => {
    switch (activeTab) {
      case 'content':
        return messageContent;
      case 'key':
        return messageKey;
      default:
        return JSON.stringify(headers);
    }
  };
  const handleKeyTabClick = (e: React.MouseEvent) => {
    e.preventDefault();
    setActiveTab('key');
  };
  const handleContentTabClick = (e: React.MouseEvent) => {
    e.preventDefault();
    setActiveTab('content');
  };
  const handleHeadersTabClick = (e: React.MouseEvent) => {
    e.preventDefault();
    setActiveTab('headers');
  };
  const keySize = new TextEncoder().encode(messageKey).length;
  const contentSize = new TextEncoder().encode(messageContent).length;
  const isContentJson = () => {
    try {
      return isObject(messageContent && JSON.parse(messageContent));
    } catch {
      return false;
    }
  };
  const isKeyJson = () => {
    try {
      return isObject(messageKey && JSON.parse(messageKey));
    } catch {
      return false;
    }
  };
  return (
    <MessageContentWrapper>
      <td colSpan={5}>
        <ContentBox>
          <SecondaryTabsStyles>
            <button
              type="button"
              className={activeTab === 'key' ? 'is-active' : ''}
              onClick={handleKeyTabClick}
            >
              Key
            </button>
            <button
              className={activeTab === 'content' ? 'is-active' : ''}
              type="button"
              onClick={handleContentTabClick}
            >
              Content
            </button>
            <button
              className={activeTab === 'headers' ? 'is-active' : ''}
              type="button"
              onClick={handleHeadersTabClick}
            >
              Headers
            </button>
          </SecondaryTabsStyles>
          <JSONViewerWrapper>
            <JSONViewer data={activeTabContent() || ''} />
          </JSONViewerWrapper>
        </ContentBox>
      </td>
      <td colSpan={2}>
        <MetadataWrapper>
          <Metadata>
            <MetadataLabel>Timestamp</MetadataLabel>
            <span>
              <MetadataValue>{timestamp?.toLocaleString()}</MetadataValue>
              <MetadataMeta>Timestamp type: {timestampType}</MetadataMeta>
            </span>
          </Metadata>

          <Metadata>
            <MetadataLabel>Content</MetadataLabel>
            <span>
              <MetadataValue>{isContentJson() ? 'JSON' : 'Text'}</MetadataValue>
              <MetadataMeta>
                Size: <BytesFormatted value={contentSize} />
              </MetadataMeta>
            </span>
          </Metadata>

          <Metadata>
            <MetadataLabel>Key</MetadataLabel>
            <span>
              <MetadataValue>{isKeyJson() ? 'JSON' : 'Text'}</MetadataValue>
              <MetadataMeta>
                Size: <BytesFormatted value={keySize} />
              </MetadataMeta>
            </span>
          </Metadata>
        </MetadataWrapper>
      </td>
    </MessageContentWrapper>
  );
};

export default MessageContent;
