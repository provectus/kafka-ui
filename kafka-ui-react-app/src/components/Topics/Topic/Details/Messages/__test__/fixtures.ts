import { TopicMessage } from 'generated-sources';

export const messages: TopicMessage[] = [
  {
    partition: 1,
    offset: 2,
    timestamp: new Date(Date.UTC(1995, 5, 5)),
    content: {
      foo: 'bar',
      key: 'val',
    },
  },
  {
    partition: 2,
    offset: 20,
    timestamp: new Date(Date.UTC(2020, 7, 5)),
    content: undefined,
  },
];

export const messageContent = `{
  "_id": "609fab8aed527f514f4e648d",
  "name": "in nostrud",
  "desc": "Dolore nostrud commodo magna velit ut magna voluptate sint aute. Excepteur aute culpa culpa dolor ipsum. Tempor est ut officia tempor laborum consectetur.\r\nAmet officia eu veniam Lorem enim aliqua aute voluptate elit do sunt in magna occaecat. Nisi sit non est adipisicing adipisicing consequat duis duis tempor consequat deserunt ea quis ad. Veniam sunt culpa nostrud adipisicing cillum voluptate non est cupidatat. Eiusmod tempor officia irure et deserunt est ex laboris occaecat adipisicing occaecat in aliquip aliqua. Do laboris culpa cupidatat cillum non. Ullamco excepteur mollit voluptate anim in nisi anim elit culpa aute. Ad officia sunt proident ut ullamco officia ea fugiat culpa cillum et fugiat aliquip.\r\nAmet non labore anim in ipsum. Et Lorem velit dolor ipsum. Irure id proident excepteur aliquip deserunt id officia dolor deserunt amet in sint. Aute in nostrud nulla ut laboris Lorem commodo nulla ipsum. Aliqua nulla commodo Lorem labore magna esse proident id ea in pariatur consectetur sint Lorem.\r\nCupidatat deserunt mollit tempor aliqua. Fugiat ullamco magna pariatur quis nulla magna. Esse duis labore ipsum nisi ullamco qui aute duis duis amet est laborum adipisicing magna. Est aliquip quis qui do aliquip nisi elit tempor ex aliquip. Excepteur aliquip ea deserunt amet adipisicing voluptate eiusmod sit sint exercitation exercitation. Id labore amet mollit ex commodo. Proident ex adipisicing deserunt esse Lorem tempor laborum nostrud commodo incididunt ea id.\r\n",
  "semster": "spring19",
  "profile": "cs",
  "degree": "bachelor",
  "degreee": "master",
  "degreeeee": "bachelor"
}`;
