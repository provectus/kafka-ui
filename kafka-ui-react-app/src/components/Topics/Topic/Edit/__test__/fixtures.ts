import { CleanUpPolicy, ConfigSource, TopicConfig } from 'generated-sources';
import { TopicWithDetailedInfo } from 'redux/interfaces/topic';

export const clusterName = 'testCluster';
export const topicName = 'testTopic';

export const config: TopicConfig[] = [
  {
    name: 'compression.type',
    value: 'producer',
    defaultValue: 'producer',
    source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'compression.type',
        value: 'producer',
        source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
      },
      {
        name: 'compression.type',
        value: 'producer',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'confluent.value.schema.validation',
    value: 'false',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [],
  },
  {
    name: 'leader.replication.throttled.replicas',
    value: '',
    defaultValue: '',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [],
  },
  {
    name: 'confluent.key.subject.name.strategy',
    value: 'io.confluent.kafka.serializers.subject.TopicNameStrategy',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [],
  },
  {
    name: 'message.downconversion.enable',
    value: 'true',
    defaultValue: 'true',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.message.downconversion.enable',
        value: 'true',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'min.insync.replicas',
    value: '1',
    defaultValue: '1',
    source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'min.insync.replicas',
        value: '1',
        source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
      },
      {
        name: 'min.insync.replicas',
        value: '1',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'segment.jitter.ms',
    value: '0',
    defaultValue: '0',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [],
  },
  {
    name: 'cleanup.policy',
    value: 'delete',
    defaultValue: 'delete',
    source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'cleanup.policy',
        value: 'delete',
        source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
      },
      {
        name: 'log.cleanup.policy',
        value: 'delete',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'flush.ms',
    value: '9223372036854775807',
    defaultValue: '9223372036854775807',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [],
  },
  {
    name: 'confluent.tier.local.hotset.ms',
    value: '86400000',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'confluent.tier.local.hotset.ms',
        value: '86400000',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'follower.replication.throttled.replicas',
    value: '',
    defaultValue: '',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [],
  },
  {
    name: 'confluent.tier.local.hotset.bytes',
    value: '-1',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'confluent.tier.local.hotset.bytes',
        value: '-1',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'confluent.value.subject.name.strategy',
    value: 'io.confluent.kafka.serializers.subject.TopicNameStrategy',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [],
  },
  {
    name: 'segment.bytes',
    value: '1073741824',
    defaultValue: '1073741824',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.segment.bytes',
        value: '1073741824',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'retention.ms',
    value: '604800000',
    defaultValue: '604800000',
    source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'retention.ms',
        value: '604800000',
        source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
      },
    ],
  },
  {
    name: 'flush.messages',
    value: '9223372036854775807',
    defaultValue: '9223372036854775807',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.flush.interval.messages',
        value: '9223372036854775807',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'confluent.tier.enable',
    value: 'false',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'confluent.tier.enable',
        value: 'false',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'confluent.tier.segment.hotset.roll.min.bytes',
    value: '104857600',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'confluent.tier.segment.hotset.roll.min.bytes',
        value: '104857600',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'confluent.segment.speculative.prefetch.enable',
    value: 'false',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'confluent.segment.speculative.prefetch.enable',
        value: 'false',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'message.format.version',
    value: '2.7-IV2',
    defaultValue: '2.7-IV2',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.message.format.version',
        value: '2.7-IV2',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'max.compaction.lag.ms',
    value: '9223372036854775807',
    defaultValue: '9223372036854775807',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.cleaner.max.compaction.lag.ms',
        value: '9223372036854775807',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'file.delete.delay.ms',
    value: '60000',
    defaultValue: '60000',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.segment.delete.delay.ms',
        value: '60000',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'max.message.bytes',
    value: '1000012',
    defaultValue: '1000012',
    source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'max.message.bytes',
        value: '1000012',
        source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
      },
      {
        name: 'message.max.bytes',
        value: '1048588',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'min.compaction.lag.ms',
    value: '0',
    defaultValue: '0',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.cleaner.min.compaction.lag.ms',
        value: '0',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'message.timestamp.type',
    value: 'CreateTime',
    defaultValue: 'CreateTime',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.message.timestamp.type',
        value: 'CreateTime',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'preallocate',
    value: 'false',
    defaultValue: 'false',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.preallocate',
        value: 'false',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'confluent.placement.constraints',
    value: '',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [],
  },
  {
    name: 'min.cleanable.dirty.ratio',
    value: '0.5',
    defaultValue: '0.5',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.cleaner.min.cleanable.ratio',
        value: '0.5',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'index.interval.bytes',
    value: '4096',
    defaultValue: '4096',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.index.interval.bytes',
        value: '4096',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'unclean.leader.election.enable',
    value: 'false',
    defaultValue: 'false',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'unclean.leader.election.enable',
        value: 'false',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'retention.bytes',
    value: '-1',
    defaultValue: '-1',
    source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'retention.bytes',
        value: '-1',
        source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
      },
      {
        name: 'log.retention.bytes',
        value: '-1',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'delete.retention.ms',
    value: '86400001',
    defaultValue: '86400000',
    source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'delete.retention.ms',
        value: '86400001',
        source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
      },
      {
        name: 'log.cleaner.delete.retention.ms',
        value: '86400000',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'confluent.prefer.tier.fetch.ms',
    value: '-1',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'confluent.prefer.tier.fetch.ms',
        value: '-1',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'confluent.key.schema.validation',
    value: 'false',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [],
  },
  {
    name: 'segment.ms',
    value: '604800000',
    defaultValue: '604800000',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [],
  },
  {
    name: 'message.timestamp.difference.max.ms',
    value: '9223372036854775807',
    defaultValue: '9223372036854775807',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.message.timestamp.difference.max.ms',
        value: '9223372036854775807',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'segment.index.bytes',
    value: '10485760',
    defaultValue: '10485760',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.index.size.max.bytes',
        value: '10485760',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
];

export const partitions = [
  {
    partition: 0,
    leader: 2,
    replicas: [
      {
        broker: 2,
        leader: false,
        inSync: true,
      },
    ],
    offsetMax: 0,
    offsetMin: 0,
  },
];

export const topicWithInfo: TopicWithDetailedInfo = {
  name: topicName,
  internal: false,
  partitionCount: 1,
  replicationFactor: 1,
  replicas: 1,
  inSyncReplicas: 1,
  segmentSize: 0,
  segmentCount: 1,
  underReplicatedPartitions: 0,
  cleanUpPolicy: CleanUpPolicy.DELETE,
  partitions,
  config,
};
