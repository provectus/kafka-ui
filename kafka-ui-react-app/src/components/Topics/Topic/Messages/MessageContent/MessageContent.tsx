import React from 'react';
import EditorViewer from 'components/common/EditorViewer/EditorViewer';
import BytesFormatted from 'components/common/BytesFormatted/BytesFormatted';
import { SchemaType, TopicMessageTimestampTypeEnum } from 'generated-sources';
import { formatTimestamp } from 'lib/dateTimeHelpers';
import { useSearchParams } from 'react-router-dom';

import * as S from './MessageContent.styled';

type Tab = 'key' | 'content' | 'headers';

export interface MessageContentProps {
  messageKey?: string;
  messageContent?: string;
  headers?: { [key: string]: string | undefined };
  timestamp?: Date;
  timestampType?: TopicMessageTimestampTypeEnum;
}

const MessageContent: React.FC<MessageContentProps> = ({
  messageKey,
  messageContent,
  headers,
  timestamp,
  timestampType,
}) => {
  const [activeTab, setActiveTab] = React.useState<Tab>('content');
  const [searchParams] = useSearchParams();
  const keyFormat = searchParams.get('keySerde') || '';
  const valueFormat = searchParams.get('valueSerde') || '';

  const activeTabContent = () => {
    switch (activeTab) {
      case 'content':
        return messageContent;
      case 'key':
        return messageKey;
      default:
        return JSON.stringify(headers);
    }
  };

  const handleKeyTabClick = (e: React.MouseEvent) => {
    e.preventDefault();
    setActiveTab('key');
  };

  const handleContentTabClick = (e: React.MouseEvent) => {
    e.preventDefault();
    setActiveTab('content');
  };

  const handleHeadersTabClick = (e: React.MouseEvent) => {
    e.preventDefault();
    setActiveTab('headers');
  };
  const keySize = new TextEncoder().encode(messageKey).length;
  const contentSize = new TextEncoder().encode(messageContent).length;
  const contentType =
    messageContent && messageContent.trim().startsWith('{')
      ? SchemaType.JSON
      : SchemaType.PROTOBUF;

  return (
    <S.Wrapper>
      <td colSpan={10}>
        <S.Section>
          <S.ContentBox>
            <S.Tabs>
              <S.Tab
                type="button"
                $active={activeTab === 'key'}
                onClick={handleKeyTabClick}
              >
                Key
              </S.Tab>
              <S.Tab
                $active={activeTab === 'content'}
                type="button"
                onClick={handleContentTabClick}
              >
                Value
              </S.Tab>
              <S.Tab
                $active={activeTab === 'headers'}
                type="button"
                onClick={handleHeadersTabClick}
              >
                Headers
              </S.Tab>
            </S.Tabs>
            <EditorViewer
              data={activeTabContent() || ''}
              maxLines={28}
              schemaType={contentType}
            />
          </S.ContentBox>
          <S.MetadataWrapper>
            <S.Metadata>
              <S.MetadataLabel>Timestamp</S.MetadataLabel>
              <span>
                <S.MetadataValue>{formatTimestamp(timestamp)}</S.MetadataValue>
                <S.MetadataMeta>Timestamp type: {timestampType}</S.MetadataMeta>
              </span>
            </S.Metadata>

            <S.Metadata>
              <S.MetadataLabel>Key Serde</S.MetadataLabel>
              <span>
                <S.MetadataValue>{keyFormat}</S.MetadataValue>
                <S.MetadataMeta>
                  Size: <BytesFormatted value={keySize} />
                </S.MetadataMeta>
              </span>
            </S.Metadata>

            <S.Metadata>
              <S.MetadataLabel>Value Serde</S.MetadataLabel>
              <span>
                <S.MetadataValue>{valueFormat}</S.MetadataValue>
                <S.MetadataMeta>
                  Size: <BytesFormatted value={contentSize} />
                </S.MetadataMeta>
              </span>
            </S.Metadata>
          </S.MetadataWrapper>
        </S.Section>
      </td>
    </S.Wrapper>
  );
};

export default MessageContent;
