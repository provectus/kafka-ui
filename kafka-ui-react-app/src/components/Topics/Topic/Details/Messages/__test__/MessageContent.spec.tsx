import { shallow } from 'enzyme';
import React from 'react';
import MessageContent from 'components/Topics/Topic/Details/Messages/MessageContent';

import { messageContent } from './fixtures';

describe('MessageContent', () => {
  const component = shallow(<MessageContent message={messageContent} />);
  describe('when it is folded', () => {
    it('matches the snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('when it is unfolded', () => {
    it('matches the snapshot', () => {
      component.find('button').simulate('click');
      expect(component).toMatchSnapshot();
    });
  });
});
