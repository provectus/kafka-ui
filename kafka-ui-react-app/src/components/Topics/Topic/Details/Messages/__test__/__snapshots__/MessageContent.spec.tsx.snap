// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageContent when it is folded matches the snapshot 1`] = `
<div
  className="is-flex is-flex-direction-column is-align-items-center"
>
  <p
    style={
      Object {
        "WebkitBackgroundClip": "text",
        "WebkitTextFillColor": "transparent",
        "background": "-webkit-linear-gradient(90deg, rgba(0, 0, 0, .1) 0%, rgba(0, 0, 0, 1) 40%)",
        "maxHeight": "160px",
        "overflow": "hidden",
        "textOverflow": "elipsis",
      }
    }
  >
    {
  "_id": "609fab8aed527f514f4e648d",
  "name": "in nostrud",
  "desc": "Dolore nostrud commodo magna velit ut magna voluptate sint aute. Excepteur aute culpa culpa dolor ipsum. Tempor est ut officia tempor laborum consectetur.
Amet officia eu veniam Lorem enim aliqua aute voluptate elit do sunt in magna occaecat. Nisi sit non est adipisicing adipisicing consequat duis duis tempor consequat deserunt ea quis ad. Veniam sunt culpa nostrud adipisicing cillum voluptate non est cupidatat. Eiusmod tempor officia irure et deserunt est ex laboris occaecat adipisicing occaecat in aliquip aliqua. Do laboris culpa cupidatat cillum non. Ullamco excepteur mollit voluptate anim in nisi anim elit culpa aute. Ad officia sunt proident ut ullamco officia ea fugiat culpa cillum et fugiat aliquip.
Amet non labore anim in ipsum. Et Lorem velit dolor ipsum. Irure id proident excepteur aliquip deserunt id officia dolor deserunt amet in sint. Aute in nostrud nulla ut laboris Lorem commodo nulla ipsum. Aliqua nulla commodo Lorem labore magna esse proident id ea in pariatur consectetur sint Lorem.
Cupidatat deserunt mollit tempor aliqua. Fugiat ullamco magna pariatur quis nulla magna. Esse duis labore ipsum nisi ullamco qui aute duis duis amet est laborum adipisicing magna. Est aliquip quis qui do aliquip nisi elit tempor ex aliquip. Excepteur aliquip ea deserunt amet adipisicing voluptate eiusmod sit sint exercitation exercitation. Id labore amet mollit ex commodo. Proident ex adipisicing deserunt esse Lorem tempor laborum nostrud commodo incididunt ea id.
",
  "semster": "spring19",
  "profile": "cs",
  "degree": "bachelor",
  "degreee": "master",
  "degreeeee": "bachelor"
}
  </p>
  <button
    className="button is-small mt-2"
    onClick={[Function]}
    title="Expand to JSON"
    type="button"
  >
    <span
      className="icon is-small"
    >
      <i
        className="fas fa-chevron-down"
      />
    </span>
  </button>
</div>
`;

exports[`MessageContent when it is unfolded matches the snapshot 1`] = `
<div
  className="is-flex is-flex-direction-column is-align-items-center"
>
  <JSONTree
    collectionLimit={50}
    data="{
  \\"_id\\": \\"609fab8aed527f514f4e648d\\",
  \\"name\\": \\"in nostrud\\",
  \\"desc\\": \\"Dolore nostrud commodo magna velit ut magna voluptate sint aute. Excepteur aute culpa culpa dolor ipsum. Tempor est ut officia tempor laborum consectetur.
Amet officia eu veniam Lorem enim aliqua aute voluptate elit do sunt in magna occaecat. Nisi sit non est adipisicing adipisicing consequat duis duis tempor consequat deserunt ea quis ad. Veniam sunt culpa nostrud adipisicing cillum voluptate non est cupidatat. Eiusmod tempor officia irure et deserunt est ex laboris occaecat adipisicing occaecat in aliquip aliqua. Do laboris culpa cupidatat cillum non. Ullamco excepteur mollit voluptate anim in nisi anim elit culpa aute. Ad officia sunt proident ut ullamco officia ea fugiat culpa cillum et fugiat aliquip.
Amet non labore anim in ipsum. Et Lorem velit dolor ipsum. Irure id proident excepteur aliquip deserunt id officia dolor deserunt amet in sint. Aute in nostrud nulla ut laboris Lorem commodo nulla ipsum. Aliqua nulla commodo Lorem labore magna esse proident id ea in pariatur consectetur sint Lorem.
Cupidatat deserunt mollit tempor aliqua. Fugiat ullamco magna pariatur quis nulla magna. Esse duis labore ipsum nisi ullamco qui aute duis duis amet est laborum adipisicing magna. Est aliquip quis qui do aliquip nisi elit tempor ex aliquip. Excepteur aliquip ea deserunt amet adipisicing voluptate eiusmod sit sint exercitation exercitation. Id labore amet mollit ex commodo. Proident ex adipisicing deserunt esse Lorem tempor laborum nostrud commodo incididunt ea id.
\\",
  \\"semster\\": \\"spring19\\",
  \\"profile\\": \\"cs\\",
  \\"degree\\": \\"bachelor\\",
  \\"degreee\\": \\"master\\",
  \\"degreeeee\\": \\"bachelor\\"
}"
    getItemString={[Function]}
    hideRoot={true}
    invertTheme={true}
    isCustomNode={[Function]}
    keyPath={
      Array [
        "root",
      ]
    }
    labelRenderer={[Function]}
    postprocessValue={[Function]}
    shouldExpandNode={[Function]}
    theme={
      Object {
        "author": "seth wright (http://sethawright.com)",
        "base00": "#1d1f21",
        "base01": "#282a2e",
        "base02": "#373b41",
        "base03": "#969896",
        "base04": "#b4b7b4",
        "base05": "#c5c8c6",
        "base06": "#e0e0e0",
        "base07": "#ffffff",
        "base08": "#CC342B",
        "base09": "#F96A38",
        "base0A": "#FBA922",
        "base0B": "#198844",
        "base0C": "#3971ED",
        "base0D": "#3971ED",
        "base0E": "#A36AC7",
        "base0F": "#3971ED",
        "scheme": "google",
      }
    }
    valueRenderer={[Function]}
  />
</div>
`;
